### PR TITLE
34 manual token labeling

### LINKED ISSUE #34: manual token labeling
@jaidhyani has more context

### PR DESCRIPTION
**Pull request by joshuawe**
Added a *.CSV file that is cerated during the labelling process of tokens.
While the pickle file (which still exists) has the structure of `dict[token_id, dict[label_name, label_value]]` the CSV now has the form

| token_id | Starts with space | Capitalized | Is Adjective | Is Adposition | ... |
|--------|--------|--------|--------|--------|--------|
| 0 | False | True| False | False | ... |
| 1 | False | False | True| False | ... |
| 2 | False | False | False | False |  ... |
| ... | ... | ... | ... | ... | ... | ... | ... |

### CODE REVIEW

File: scripts/label_all_tokens.py
Comment (jaidhyani) [inline scripts/label_all_tokens.py]: We should probably make the path a script argument, and not default to outputting data files in code. This isn't a blocker though, we can adjust it later.

**Comment by jaidhyani** [comment]
This is looking good. Here's what I think we want to do to finalize it:

1. Factor out the core token-labeling functionality into the `token_labelling.py` (basically lines 49-88). This probably looks like a function that takes a tokenizer as an argument and returns a `labelled_token_ids_dict` (or similar data structure that fully specifies token labels)

2. Adapt the current script into an `export_labels` script with three arguments: (1) model/tokenizer (as exists in the current script) (2) output format (csv or pkl) and (3) output path. Output path should be a mandatory argument without defaults. This will call the function defined in (1) and basically just handle the file writing.

3. Back in `token_labeling.py`, add a function `import_token_labels` which takes a path argument (to a csv) and returns a `labelled_token_ids_dict`-like data structure specifying token labels.

4. Move generated data files to `delphi/static/` (see currently-a-draft PR #41)

**Comment by jettjaniak** [comment]
Nice! The CSV needs a column with token label though

File: scripts/label_all_tokens.py
Comment (jettjaniak) [inline scripts/label_all_tokens.py]:
```suggestion
        "--save-dir", type=str, help="Directory to save the results.", required=True
```

File: scripts/label_all_tokens.py
Comment (jettjaniak) [inline scripts/label_all_tokens.py]:
```suggestion
        "--output-format",
```

**Comment by joshuawe** [comment]
> Nice! The CSV needs a column with token label though

@jettjaniak Could you elaborate what you mean?
Would you like to have a header line, that explains the meaning of each column?

**Comment by jaidhyani** [comment]
This is looking good - nice.

File: src/delphi/eval/token_labelling.py
Comment (jettjaniak) [inline src/delphi/eval/token_labelling.py]: why is this global?

File: tests/eval/test_token_labelling.py
Comment (jettjaniak) [inline tests/eval/test_token_labelling.py]: what does it do?

File: scripts/label_all_tokens.py
Comment (jettjaniak) [inline scripts/label_all_tokens.py]: uppercase is generally reserved for constants

**Comment by jettjaniak** [comment]
I don't see a use case for *.pkl anymore

File: tests/eval/test_token_labelling.py
Comment (joshuawe) [reply tests/eval/test_token_labelling.py]: This decorator has another test run first (the test that it depends on), whose results can be used in another test.
So here, we first create a dict with the token labels in another test and use those results to test subsequent functions of our library.

File: src/delphi/eval/token_labelling.py
Comment (joshuawe) [reply src/delphi/eval/token_labelling.py]: thanks for catching that. Needed the global var during testing, not in the library code

File: scripts/label_all_tokens.py
Comment (joshuawe) [reply scripts/label_all_tokens.py]: made it lower case

**Comment by joshuawe** [comment]
Resolved two comments of yours.

Should I get rid of the *.pkl option entirely before merging? @jettjaniak

File: tests/eval/test_token_labelling.py
Comment (jettjaniak) [inline tests/eval/test_token_labelling.py:146]: beartype is doing this automatically

File: tests/eval/test_token_labelling.py
Comment (jettjaniak) [reply tests/eval/test_token_labelling.py]: I see. I don't think this is a good practice, I suggest you use pytest fixtures instead (built-in, don't require additional packages) https://docs.pytest.org/en/6.2.x/fixture.html

**Comment by jettjaniak** [comment]
I think #47 and #51 should go with this PR
And please add the CSV to static when ready

File: tests/eval/test_token_labelling.py
Comment (joshuawe) [reply tests/eval/test_token_labelling.py]: In general I disagree about good practice. In bigger projects with large amounts of tests, you would not want to waste compute on tests that anyways depend on other tests. 

Here we do not have a large test set, so if you prefer I can go with fixtures and have the same code from `test_label_tokens_from_tokenizer` a second time.

File: tests/eval/test_token_labelling.py
Comment (joshuawe) [reply tests/eval/test_token_labelling.py:146]: Ha, nice. But weirdly enough I needed this to catch a bug because beartype did not throw an error when the type was `dict[int, dict[str, np.array[bool]]`

**Comment by joshuawe** [comment]
- [x] get rid of pickling
- [ ] add str repr into *.csv
  - [x] correctly create & save CSV
  - [ ] correctly load the CSV
  - [ ] Merge the code into this branch of #40
- [x] rewrite testing to not include dependency (slow version OR @pytest.skip)
- [ ] Add the token labelling CSV once it is created to STATIC
- [ ] comment / todo / fixme on export to csv eating spaces

**Comment by jettjaniak** [comment]
- check what's wrong with beartype and nested structures https://beartype.readthedocs.io/en/latest/eli5/
- implement slow version of tests and @pytest.skip or do fixture
- comment / todo / fixme on export to csv eating spaces

**Comment by jettjaniak** [comment]
@transcendingvictor your tasks is blocked by this and Josh won't be available until weekend. Could you complete the items in the comment above and merge this? Hit me up on Discord if sth is unclear

File: tests/eval/test_token_labelling.py
Comment (jettjaniak) [inline tests/eval/test_token_labelling.py:1]: this file should be renamed to spacy_* as well

File: tests/eval/test_token_labelling.py
Comment (jettjaniak) [inline tests/eval/test_token_labelling.py]: How did you rename the files? If you do right click and rename in VSCode it should also rename all references

File: tests/eval/test_token_labelling.py
Comment (jettjaniak) [inline tests/eval/test_token_labelling.py]: this is a fixture, not a test, so you probably don't have to skip it

File: tests/eval/test_token_labelling.py
Comment (jettjaniak) [inline tests/eval/test_token_labelling.py]: I didn't realize how many there are and that they're all in a single file. I believe you can just do pytest.skip() on the top level, just after the imports to skip the whole file. The reason is "tests are slow and we're not using this module currently"

**Review by jettjaniak — APPROVED** [review]

**Comment by joshuawe** [comment]
Thanks @transcendingvictor  for doing this

**Pull request merged**
